internal/build/imgsrc/depot.go

```diff
@@ -0,0 +1,331 @@
+package imgsrc
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"time"
+
+	depotbuild "github.com/depot/depot-go/build"
+	depotmachine "github.com/depot/depot-go/machine"
+	"github.com/moby/buildkit/client"
+	"github.com/moby/buildkit/session/secrets/secretsprovider"
+	"github.com/moby/buildkit/util/progress/progressui"
+	"github.com/pkg/errors"
+	"github.com/superfly/fly-go"
+	"github.com/superfly/flyctl/helpers"
+	"github.com/superfly/flyctl/internal/cmdfmt"
+	"github.com/superfly/flyctl/internal/config"
+	"github.com/superfly/flyctl/internal/flyutil"
+	"github.com/superfly/flyctl/internal/metrics"
+	"github.com/superfly/flyctl/internal/render"
+	"github.com/superfly/flyctl/internal/tracing"
+	"github.com/superfly/flyctl/iostreams"
+	"github.com/superfly/flyctl/terminal"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+	"golang.org/x/sync/errgroup"
+)
+
+var _ imageBuilder = (*DepotBuilder)(nil)
+
+type DepotBuilder struct{}
+
+func (d *DepotBuilder) Name() string { return "depot.dev" }
+
+func (d *DepotBuilder) Run(ctx context.Context, _ *dockerClientFactory, streams *iostreams.IOStreams, opts ImageOptions, build *build) (*DeploymentImage, string, error) {
+	ctx, span := tracing.GetTracer().Start(ctx, "depot_builder", trace.WithAttributes(opts.ToSpanAttributes()...))
+	defer span.End()
+
+	build.BuildStart()
+
+	var dockerfile string
+
+	if opts.DockerfilePath != "" {
+		if !helpers.FileExists(opts.DockerfilePath) {
+			build.BuildFinish()
+			err := fmt.Errorf("dockerfile '%s' not found", opts.DockerfilePath)
+			tracing.RecordError(span, err, "failed to find dockerfile")
+			return nil, "", err
+		}
+		dockerfile = opts.DockerfilePath
+	} else {
+		dockerfile = ResolveDockerfile(opts.WorkingDir)
+	}
+
+	if dockerfile == "" {
+		span.AddEvent("dockerfile not found, skipping")
+		terminal.Debug("dockerfile not found, skipping")
+		build.BuildFinish()
+		return nil, "", nil
+	}
+
+	var relDockerfile string
+	if isPathInRoot(dockerfile, opts.WorkingDir) {
+		// pass the relative path to Dockerfile within the context
+		p, err := filepath.Rel(opts.WorkingDir, dockerfile)
+		if err != nil {
+			tracing.RecordError(span, err, "failed to get relative dockerfile path")
+			build.BuildFinish()
+			return nil, "", err
+		}
+		// On Windows, convert \ to a slash / as the docker build will
+		// run in a Linux VM at the end.
+		relDockerfile = filepath.ToSlash(p)
+	}
+	span.SetAttributes(attribute.String("relative_dockerfile_path", relDockerfile))
+
+	build.ImageBuildStart()
+
+	image, err := depotBuild(ctx, streams, opts, dockerfile, build)
```

Comment:
Is there a reason `depotBuild` can't pull `streams *iostreams.IOStreams` out of the context object on its own? Cause I think it makes the function signature easier to read and the remaining args coherent.

Reply:
Cool, I don't have any strong feelings about it